skill_predictor.py

```diff
@@ -0,0 +1,210 @@
+            log += 1
+            entries = list(time_stamps.keys())
+            if len(entries) <= 1:
+                continue
+            for i in range(1, len(entries)):
+                curr_skills = time_stamps[entries[i]]
+                prev_skills = time_stamps[entries[i-1]]
+                if len(prev_skills) <= 2:
+                    continue
+
+                # bag transition size
+                if len(curr_skills) != len(prev_skills):
+                    if len(prev_skills) not in self.mle_bagsize:
+                        self.mle_bagsize[len(prev_skills)
+                                         ] = collections.defaultdict(int)
+                    self.mle_bagsize[len(prev_skills)][len(curr_skills)] += 1
+
+                # skill additions
+                for skill in curr_skills:
+                    if skill not in prev_skills:
+                        with open('train_data.txt', 'a') as f, open('test_data.txt', 'a') as g:
+                            counter = 0
+                            tmp_prev_skills = itertools.permutations(
+                                list(prev_skills))
+                            for permutation in tmp_prev_skills:
+                                curr_sentence = list(permutation) + [skill]
+                                curr_sentence = ' '.join(curr_sentence)
+                                if np.random.rand() < 0.8:
+                                    f.write('%s\n' % curr_sentence)
+                                else:
+                                    g.write('%s\n' % curr_sentence)
+                                counter += 1
+                                if counter > num_permutations:
+                                    break
+
+                # skill deletions
+                for skill in prev_skills:
+                    if skill not in curr_skills:
+                        with open('train_data.txt', 'a') as f, open('test_data.txt', 'a') as g:
+                            counter = 0
+                            tmp_prev_skills = itertools.permutations(
+                                list(prev_skills))
+                            for permutation in tmp_prev_skills:
+                                curr_sentence = list(
+                                    permutation) + ['-' + skill]
+                                curr_sentence = ' '.join(curr_sentence)
+                                if np.random.rand() < 0.8:
+                                    f.write('%s\n' % curr_sentence)
+                                else:
+                                    g.write('%s\n' % curr_sentence)
+                                counter += 1
+                                if counter > num_permutations:
+                                    break
+
+    def train(self):
+        """
+        Trains the given KenLM model on a language model dataset
+        """
+        try:
+            self.model = kenlm.Model('../kenlm/build/train_data.arpa')
```

Comment:
Don't put dependencies outside of the current directory branch. You need to guarantee kenlm is available, and unless you can enforce its installation and its installation in a particular directory, then it should be provided locally to your script (e.g. '/kenlm/' or similar).